src/ai/backend/manager/api/session.py

```diff
@@ -1869,25 +1871,59 @@ async def destroy(request: web.Request, params: Any) -> web.Response:
     #         not request['is_superadmin'] and request['is_admin']:
     #     domain_name = request['user']['domain_name']
     log.info(
-        "DESTROY (ak:{0}/{1}, s:{2}, forced:{3})",
+        "DESTROY (ak:{0}/{1}, s:{2}, forced:{3}, recursive: {4})",
         requester_access_key,
         owner_access_key,
         session_name,
         params["forced"],
+        params["recursive"],
     )
-    last_stat = await root_ctx.registry.destroy_session(
-        functools.partial(
-            root_ctx.registry.get_session,
-            session_name,
-            owner_access_key,
-            # domain_name=domain_name,
-        ),
-        forced=params["forced"],
-    )
-    resp = {
-        "stats": last_stat,
-    }
-    return web.json_response(resp, status=200)
+
+    requester_access_key, owner_access_key = await get_access_key_scopes(request)
+
+    if params["recursive"]:
+        async with root_ctx.db.begin_readonly() as conn:
+            dependent_session_ids = await find_dependent_sessions(
+                session_name, conn, owner_access_key
+            )
+
+            last_stats = await asyncio.gather(
+                *map(
+                    lambda session_id_or_name: root_ctx.registry.destroy_session(
+                        functools.partial(
+                            root_ctx.registry.get_session,
+                            session_id_or_name,
+                            owner_access_key,
+                            # domain_name=domain_name,
+                        ),
+                        forced=params["forced"],
+                    ),
+                    [*dependent_session_ids, session_name],
+                ),
+                return_exceptions=True,
+            )
```

Comment:
This could be refactored as:
```suggestion
            last_stats = await asyncio.gather(
                *[
                    root_ctx.registry.destroy_session(
                        functools.partial(
                            root_ctx.registry.get_session,
                            session_id_or_name,
                            owner_access_key,
                            # domain_name=domain_name,
                        ),
                        forced=params["forced"],
                    )
                    for session_id_or_name in
                    [*dependent_session_ids, session_name],
                ],
                return_exceptions=True,
            )
```